OSS layer of the session summarizer

bl-nero opened the pull request:
- The backend service
- A stub gRPC service which communicates that the feature is only available in Enterprise
- e_imports statement that adds an OpenAI client that will be used by the enterprise build
- Necessary wiring that allows `ProtoStream` instances to notify listeners about upload completion; the accompanying enterprise change will use it to trigger summarization

RFD: https://github.com/gravitational/teleport.e/blob/master/rfd/0026e-ai-session-summaries.md
Contributes to https://github.com/gravitational/teleport/issues/47083
Requires https://github.com/gravitational/teleport/pull/56377

bl-nero committed:
OSS layer of the session summarizer

- The backend service
- A stub gRPC service which communicates that the feature is only
  available in Enterprise
- e_imports statement that adds an OpenAI client that will be used by
  the enterprise build
- Necessary wiring that allows `ProtoStream` instances to notify
  listeners about upload completion; the accompanying enterprise change
  will use it for triggering summarization

bl-nero committed:
Configuration resources for the session summarizer

bl-nero committed:
Review

bl-nero commented on lib/auth/auth.go:
```diff
@@ -1206,6 +1216,12 @@ type Server struct {
 	// loginHooks are a list of hooks that will be called on login.
 	loginHooks []LoginHook
 
+	// uploadCompletionHooksMu is a mutex to protect the [uploadCompletionHooks].
+	uploadCompletionHooksMu sync.RWMutex
+	// uploadCompletionHooks are a list of hooks that will be called when a
+	// session recording upload completes.
+	uploadCompletionHooks []events.UploadCompletionHook
```
@rosstimothy TL;DR: yes, it's partially due to how the enterprise plugin works.

Here comes the full story. The enterprise plugin gets a fully initialized auth server and needs to augment it with its own logic. In my case, I'm registering the upload completion hook in `e/lib/auth/plugin.go`. Instead of registering it with the auth server, I could register it with the streamer. The problem here is that I would have to add the registration method to the `events.Streamer` interface, but that would then force me to implement these in some ridiculous places. And I don't even mean the `DiscardStreamer`; this goes as far as the `authclient.ClientI` interface by composition, so things like `authclient.Client` would be affected. The streamer is, unfortunately, the last component down the composition matryoshka where I can put something; everything down from there is ephemeral and is created on demand, and it starts with `ProtoStream`, which currently gets the completion hook on creation from its creator, the `ProtoStreamer`, in ` `ProtoStreamer.CreateAuditStreamForUpload`.

I'm open to ideas, but so far, this is the simplest way. Note that "let's call the summarizer directly from the `sliceWriter`" strategy bears the same consequences: you need to have a reference to the streamer service in the slice writer. Which means you need to push it down the chain from somewhere, and that "somewhere" happens to be the auth server — only this time, we get a similar ugly structure with additional coupling between the slice writer and the summarizer.

It's also worth noting that even if it wasn't for the enterprise-related complications, even the OSS code was quite unforgiving; I remember that when we were experimenting and I didn't even care about going enterprise, we had a chicken-and-egg problem when initializing services; something about two services depending on each other, which could only be solved by setting one of the dependencies after creating a service — or using a callback pattern like here.

r0mant commented on lib/auth/summarizer/summarizerv1/service.go:
```diff
@@ -0,0 +1,139 @@
+package summarizerv1
```
Nit: Missing license header.

r0mant commented on lib/auth/summarizer/summarizerv1/service.go:
```diff
@@ -0,0 +1,139 @@
+
+// ossSummarizerService is an OSS version of the SummarizerService. It returns
+// a licensing error from every RPC.
+type ossSummarizerService struct {
```
Nit: I think "oss" prefix is redundant here. I'd just call it `summarizerService`.

```suggestion
type summarizerService struct {
```

r0mant commented on lib/services/local/summarizer.go:
```diff
@@ -0,0 +1,203 @@
+package local
```
Missing license header.

r0mant commented on lib/services/presets.go:
```diff
@@ -218,6 +218,9 @@ func NewPresetEditorRole() types.Role {
 					types.NewRule(types.KindSigstorePolicy, RW()),
 					types.NewRule(types.KindWorkloadIdentityX509IssuerOverride, RW()),
 					types.NewRule(types.KindWorkloadIdentityX509IssuerOverrideCSR, RW()),
+					types.NewRule(types.KindSummarizationInferenceModel, RW()),
+					types.NewRule(types.KindSummarizationInferenceSecret, RW()),
+					types.NewRule(types.KindSummarizationInferencePolicy, RW()),
```
Are these the actual yaml configuration resources a user is supposed to be able to create/edit? I don't see their protos in this PR. And also if so, why do we have 3 new resources, can't we have a single "summarization inference config" or something to simplify?

bl-nero commented on lib/auth/auth.go:
```diff
@@ -1338,6 +1354,42 @@ func (a *Server) ResetLoginHooks() {
+
+	var errs []error
+	for _, hook := range completionHooks {
+		errs = append(errs, trace.Wrap(hook(ctx, sessionID, sessionEndEvent)))
```
I'd say yes, in this pattern, I'd be surprised if one hook's behavior influenced whether another one gets called. I'll add it to the docs.

bl-nero committed:
review

bl-nero commented on lib/events/stream.go:
```diff
@@ -648,6 +670,18 @@ func (w *sliceWriter) receiveAndUpload() error {
 
 				continue
 			}
+			// Capture the session end event. Note that we deliberately support more
```
Yes, I tested both modes.

bl-nero commented on lib/service/service.go:
```diff
@@ -2196,6 +2197,15 @@ func (process *TeleportProcess) initAuthService() error {
 		streamer, err = events.NewProtoStreamer(events.ProtoStreamerConfig{
 			Uploader:  uploadHandler,
 			Encrypter: encryptedIO,
+			CompletionHook: func(ctx context.Context, sessionID session.ID, sessionEndEvent *apievents.OneOf) error {
```
If it feels like a circular dependency, it's because it kinda is, and it's related to [Tim's comment](https://github.com/gravitational/teleport/pull/56551#discussion_r2192957844). This basically can't be done in a trivial way, since the summarizer needs to use a streamer to get to the session data, while the streamer is also where the session upload end happens. So instead of introducing an explicit circular dependency between the summarizer and the session uploader, I decided to use a hook pattern similar to the existing login hooks. The streamer doesn't care then about the summarization, and the summarizer can freely use the streamer as a dependency.

Attempting to create an explicit circular dependency here would also run into the same obstacle that I mentioned in the thread with Tim above: this would mean that injecting the summarizer as the streamer dependency leads to adding a setter method on the streamer interface, and this in turn ends up polluting much more types in our architecture, most of them not needing this at all.

bl-nero replied:
(Implementation changed.)

bl-nero committed:
review

bl-nero committed:
Rename the resources

bl-nero committed:
Rename the resources

bl-nero committed:
Merge branch 'bl-nero/summarizer-protos' into bl-nero/summaries

bl-nero committed:
Review

bl-nero committed:
go mod tidy

bl-nero committed:
lint

bl-nero committed:
Add response messages

bl-nero committed:
Merge branch 'bl-nero/summarizer-protos' into bl-nero/summaries

bl-nero committed:
Add return types to the gRPC server

bl-nero committed:
Split out the resources gRPC service from summarizer

bl-nero committed:
Merge branch 'master' into bl-nero/summaries

bl-nero committed:
Fix CI issues

bl-nero committed:
Add some comments, fix existing ones

bl-nero committed:
Merge branch 'master' into bl-nero/summaries

bl-nero committed:
Remove unnecessary e/ change

bl-nero committed:
Fix broken lint

r0mant commented on lib/auth/summarizer/summarizerv1/summarizer.go:
```diff
@@ -0,0 +1,68 @@
+	Summarize(ctx context.Context, sessionID session.ID, sessionEndEvent *events.OneOf) error
+}
+
+// SummarizerWrapper is a wrapper around the SummarizerService interface. Its
+// purpose is to allow substituting the wrapped service after a dependent
+// service has been configured with the wrapper as the service implementation.
```
> Its purpose is to allow substituting the wrapped service after a dependent service has been configured

Why is this needed?

r0mant reviewed (commented):
Generally looks reasonable but the "summarizer wrapper" bit I find somewhat confusing. Is there a way to simplify this, replacing implementation after the service has been initialized is not a pattern I think we use much elsewhere and seems a little hacky/prone to mistakes.

bl-nero committed:
Replace summarizer wrapper with summarizer provider

This fixes a minor race condition, adds some missing unit tests, and
hopefully reduces confusion about the entire concept.

bl-nero committed:
Fix license and import order

bl-nero committed:
review

zmb3 commented on lib/events/stream.go:
```diff
@@ -771,11 +810,15 @@ func (w *sliceWriter) completeStream() {
+		}
+
+		summarizer := w.proto.cfg.SummarizerProvider.ProvideSummarizer()
+		err = summarizer.Summarize(w.proto.cancelCtx, w.proto.cfg.Upload.SessionID, w.sessionEndEvent)
```
Is this a long-running operation? If so, is it okay to make `completeStream` take longer, or should this be done in a new goroutine?

zmb3 commented on lib/events/stream_test.go:
```diff
@@ -370,6 +372,96 @@ func TestEncryptedRecordingIO(t *testing.T) {
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx, cancel := context.WithTimeout(t.Context(), 10*time.Second)
+			defer cancel()
```
If your tests don't actually depend on the 10-second timeout, you can just use `t.Context()` for a context.

zmb3 commented on lib/events/stream_test.go:
```diff
@@ -370,6 +372,96 @@ func TestEncryptedRecordingIO(t *testing.T) {
+
+			mockSummarizer.AssertExpectations(t)
+
+			// doneC := make(chan struct{})
+			// go func() {
+			// 	defer close(doneC)
+			// 	stream.Complete(ctx)
+			// 	stream.Close(ctx)
+			// }()
+
+			// select {
+			// case <-ctx.Done():
+			// 	t.Fatal("Timeout waiting for emitter to complete")
+			// case <-doneC:
+			// }
```
Remove this if we no longer need it.

bl-nero committed:
Clean up tests

bl-nero committed:
Review

Move code around, switch from atomic pointer to a mutex in the
summarizer provider.

bl-nero committed:
Fix stream_test

bl-nero commented:
Retested once again on all combinations of:

- OSS and enterprise
- OpenSSH node, external Teleport node, and a node running within the auth server.
- All recording modes (node, node-sync, proxy, proxy-sync)

bl-nero committed:
Review

bl-nero committed:
Merge branch 'master' into bl-nero/summaries

bl-nero committed:
Narrow down the number of event types accepted by Summarizer

bl-nero committed:
Fix tests

bl-nero commented on lib/auth/summarizer/summarizerv1/summarizer.go:
```diff
@@ -0,0 +1,89 @@
+	}
+
+	sptr := p.summarizer.Load()
+	if sptr == nil {
```
I tried to support the "typed interface pointing to nil value" case, but it turned out to open a new can of worms when I applied your comment about whether a pointer receiver is necessary in some dummy implementations. When you substitute a struct instead of a pointer, the `IsNil` assertion panics, because the underlying value is not a pointer. Now I could instead test whether given object is a pointer, but it drags us deeper into a rabbit hole of absurd: what if the underlying summarizer implementation is a slice, and that is also nullable? So all in all, I decided that at some point, we just have to assume that nobody is going to go to crazy lengths to break this interface and I just left it at a simple nil check that doesn't use reflection. There are literally two other places in our codebase where we care about protecting from nil values to this length, so I think it's safe to assume this will be an overkill in this case.

bl-nero commented on lib/auth/summarizer/summarizer.go:
```diff
@@ -0,0 +1,34 @@
+	// sessionEndEvent is optional, but should be specified if possible, as an
+	// optimization to skip reading the session stream in order to find the end
+	// event.
+	Summarize(ctx context.Context, sessionID session.ID, sessionEndEvent *events.OneOf) error
```
OK, I'll modify the summarizer interface. It sounds simpler than modifying that type.

bl-nero committed:
Review

bl-nero commented on lib/events/eventstest/generate.go:
```diff
@@ -148,3 +151,154 @@ func GenerateTestSession(params SessionParams) []apievents.AuditEvent {
+
+// GenerateTestDBSession generates test database session events starting with
+// session start event, adds params.Queries events and returns the result.
+func GenerateTestDBSession(params DBSessionParams) []apievents.AuditEvent {
```
Yes, I'm planning to use it in the enterprise tests. (I'm already using `GenerateTestSession`.)

bl-nero commented on lib/auth/summarizer/summarizer.go:
```diff
@@ -0,0 +1,41 @@
+	// SummarizeSSH summarizes a database session recording that ended with a given
+	// event.
+	SummarizeDatabase(ctx context.Context, sessionEndEvent *events.DatabaseSessionEnd) error
+	// SummarizeUnknown summarizes a session recording with a given ID. This is
+	// used for cases where the session ID is known, but there is no end event
+	// available. [SessionSummarizer.SummarizeSSH] and
+	// [SessionSummarizer.SummarizeDatabase] should be used instead of this
+	// method whenever possible, as they are more efficient.
+	SummarizeUnknown(ctx context.Context, sessionID session.ID) error
```
Changed to `SummarizeWithoutEndEvent`.

bl-nero committed:
Review: some renamings

rosstimothy reviewed (approved):
:shipit:
